### PR TITLE
scRNA-seq skeleton for future batch-correction/sub-clustering tutorial plug-ins

### PR DESCRIPTION
**Pull request by nomadscientist**
This tutorial is a continuation of the Alevin one, and hopefully - all tool upgrade being sorted - will allow for future connections to trajectories in scanpy/batch-correction/sub-clustering etc. But I figured trying to work on 4 at once would be a horrendous idea so one at a time...
Also a couple of edits on the Alevin tutorial

### CODE REVIEW

**Comment by nomadscientist** [comment]
Also Github is lying. Wendi does not want to merge anything yet. Wendi wants to work on this profusely for more days before considering such a thing. Shame on you Github.

**Comment by nomadscientist** [comment]
OK. It is ready for other eyes now! @bebatut @hexylena @pinin4fjords

File: topics/transcriptomics/tutorials/scrna-seq-basic-pipeline/tutorial.md [hunk 1]
```diff
@@ -0,0 +1,910 @@
...
+
+---
+
+
+# Introduction
+{:.no_toc}
+
+<!-- This is a comment. -->
+
+You've done all the work to make a single cell matrix, with gene counts and mitochondrial counts and buckets of cell metadata from all your variables of interest (or, if not, please see [this tutorial](https://training.galaxyproject.org/training-material/topics/transcriptomics/tutorials/droplet-quantification-preprocessing/tutorial.html) to do so!) Now it's time to fully process our data, to remove low quality cells, to reduce the many dimensions of data that make it difficult to work with, and ultimately to try to define our clusters and to find our biological meaning and insights! There are many packages for analysing single cell data - Seurat, Scanpy, Monocle3, Scater, and so forth. We're working with Scanpy, because currently Galaxy hosts the most Scanpy tools of all of those options.
```
Comment (mtekman) [inline topics/transcriptomics/tutorials/scrna-seq-basic-pipeline/tutorial.md:42]: maybe a citation for each of these

File: topics/transcriptomics/tutorials/scrna-seq-basic-pipeline/tutorial.md [hunk 2]
```diff
@@ -0,0 +1,910 @@
...
+> In this tutorial, we will cover:
+>
+> 1. TOC
+> {:toc}
+>
+{: .agenda}
+
+## Get data
+
+We've provided you with experimental data to analyse from a mouse dataset of fetal growth restriction {% cite Bacon2018 %}. This is the full dataset generated from [this tutorial](https://training.galaxyproject.org/training-material/topics/transcriptomics/tutorials/droplet-quantification-preprocessing/tutorial.html) if you used the full fastq files rather than the subsampled ones (see the study in Single Cell Expression Atlas [here](https://www.ebi.ac.uk/gxa/sc/experiments/E-MTAB-6945/results/tsne) and the project submission [here](https://www.ebi.ac.uk/arrayexpress/experiments/E-MTAB-6945/)). You can find this dataset in this [input history](https://humancellatlas.usegalaxy.eu/u/wendi.bacon.training/h/filter-plot-and-explore-single-cell-rna-seq-data---input) or download from Zenodo below.
```
Comment (mtekman) [inline topics/transcriptomics/tutorials/scrna-seq-basic-pipeline/tutorial.md:60]: fastq → FASTQ

File: topics/transcriptomics/tutorials/scrna-seq-basic-pipeline/tutorial.md [hunk 3]
```diff
@@ -0,0 +1,910 @@
...
+
+## Analysing the plots
+
+That's a lot of information! Let's attack this in sections and see what questions these plots can help us answer. The *scratchbook* {% icon galaxy-scratchbook %} may help here to look at the different plots at the same time!
+
+{% include snippets/use_scratchbook.md %}
+
+> ### {% icon question %} Question - Batch Variation
+>
+> Are their differences in sequencing depth across the samples?
```
Comment (mtekman) [inline topics/transcriptomics/tutorials/scrna-seq-basic-pipeline/tutorial.md:196]: their → there

File: topics/transcriptomics/tutorials/scrna-seq-basic-pipeline/tutorial.md [hunk 4]
```diff
@@ -0,0 +1,910 @@
...
+> 2. How do you interpret it?
+>
+> > ### {% icon solution %} Solution
+> >
+> > 1. The plot `violin - batch - log` will have what you're looking for!
+> >     ![Violin - batch - log](../../images/wab-violin-batch-log.png "Violin - batch - log (Raw)")
+> > 2. Keeping in mind that this is a log scale - which means that small differences can mean large differences - the violin plots probably look pretty similar.
+> >    - `N707` and `N703` might be a bit lower on genes and counts (or UMIs), but the differences aren't catastrophic.
+> >    - The `pct_counts_mito` looks pretty similar across the batches, so this also looks good.
+> >    - Nothing here would cause me to eliminate a sample from my analysis, but if you see a sample looking completely different from the rest, you would need to question why that is and consider eliminating it from your experiment!
```
Comment (mtekman) [inline topics/transcriptomics/tutorials/scrna-seq-basic-pipeline/tutorial.md:207]: me → us

File: topics/transcriptomics/tutorials/scrna-seq-basic-pipeline/tutorial.md [hunk 5]
```diff
@@ -0,0 +1,910 @@
...
+> >
+> {: .solution}
+>
+{: .question}
+
+> ### {% icon question %} Question - Biological Variables
+>
+> Are their differences in sequencing depth across sex? Genotype?
```
Comment (mtekman) [inline topics/transcriptomics/tutorials/scrna-seq-basic-pipeline/tutorial.md:215]: their → there

File: topics/transcriptomics/tutorials/scrna-seq-basic-pipeline/tutorial.md [hunk 6]
```diff
@@ -0,0 +1,910 @@
...
+> >
+> {: .solution}
+{: .question}
+
+## Apply the Thresholds
+
+It's now time to apply these thresholds to our data! First, a reminder of how many cells and genes are in your object: `25281 cells` and `35734 genes`. Let's see how that changes each time!
+
+> ### {% icon details %} Working in a group? Decision-time!
+> If you are working in a group, you can now divvy up a decision here with one *control* and the rest varied numbers so that you can compare results throughout the tutorials.
```
Comment (mtekman) [inline topics/transcriptomics/tutorials/scrna-seq-basic-pipeline/tutorial.md:293]: "divvy up" maybe simplify phrases like this, as I'm not sure how well they translate

File: topics/transcriptomics/tutorials/scrna-seq-basic-pipeline/tutorial.md [hunk 7]
```diff
@@ -0,0 +1,910 @@
...
+>
+{: .question}
+
+Here's a quick overall summary for easy visualisation if you fancy it.
+![Violinplot-Summary](../../images/wab-violins2gether.png "Filtering summary")
+
+Fantastic work! However, you've now removed a whole heap of cells, and since the captured genes are sporadic (i.e. a small percentage of the overall transcriptome per cell) this means there are a number of genes in your matrix that are currently not in any of the remaining cells. Genes that do not appear in any cell, or even in only 1 or 2 cells, will make some analytical tools break and overall will not be biologically informative. So let's remove them! Note that `3` is not necessarily the best number, rather it is a fairly conservative threshold. You could go as high as 10 or more.
+
+> ### {% icon details %} Working in a group? Decision-time!
+> If you are working in a group, you can now divvy up a decision here with one *control* and the rest varied numbers so that you can compare results throughout the tutorials.
```
Comment (mtekman) [inline topics/transcriptomics/tutorials/scrna-seq-basic-pipeline/tutorial.md:442]: "divvy up".... hmm, well I guess if you use it more than once it's okay :-)

File: topics/transcriptomics/tutorials/scrna-seq-basic-pipeline/tutorial.md [hunk 8]
```diff
@@ -0,0 +1,910 @@
...
+> - Variable: **n_cells**
+> - Control > `3`
+> - Everyone else: Choose your own thresholds and compare results! Note if you go less than 3 (or even remove this step entirely), future tools are likely to fail due to empty gene data.
+{: .details}
+
+
+> ### {% icon hands_on %} Hands-on: Filter genes
+>
+> 1. {% tool [Scanpy FilterGenes](https://humancellatlas.usegalaxy.eu/root?tool_id=toolshed.g2.bx.psu.edu/repos/ebi-gxa/scanpy_filter_genes/scanpy_filter_genes/1.6.0+galaxy0) %} with the following parameters:
```
Comment (mtekman) [inline topics/transcriptomics/tutorials/scrna-seq-basic-pipeline/tutorial.md:451]: yep agree with the bot

File: topics/transcriptomics/tutorials/scrna-seq-basic-pipeline/tutorial.md [hunk 9]
```diff
@@ -0,0 +1,910 @@
...
+
+> ### {% icon comment %} 2262???
+> Where did the `2262` come from? The quickest way to figure out how many highly variable genes you have, in my opinion, is to re-run {% icon galaxy-refresh %} the **Scanpy FindVariableGenes** tool and select the parameter to *Remove genes not marked as highly variable*. Then you can Inspect your resulting object and you'll see only 2262 genes. I prefer to keep these in the object, just in case something is important (paranoia?), but you will get the same results whether you keep these or remove them. The following steps all use only highly variable genes.
+{: .comment}
+
+> ### {% icon hands_on %} Hands-on: Calculate Principal Components
+>
+> 1. {% tool [Scanpy RunPCA](https://humancellatlas.usegalaxy.eu/root?tool_id=toolshed.g2.bx.psu.edu/repos/ebi-gxa/scanpy_run_pca/scanpy_run_pca/1.6.0+galaxy1) %} with the following parameters:
```
Comment (mtekman) [inline topics/transcriptomics/tutorials/scrna-seq-basic-pipeline/tutorial.md:527]: yep, good bot

File: topics/transcriptomics/tutorials/scrna-seq-basic-pipeline/tutorial.md [hunk 10]
```diff
@@ -0,0 +1,910 @@
...
+
+![PCA-variance](../../images/wab-pcavariance.png "Variability explained per PC")
+
+We can see that there is really not much variation explained past component 19. So we might save ourselves a great deal of time and muddied data by focusing on the top `20` PCs.
+
+## Neighborhood graph
+
+We're still looking at around 20 dimensions at this point. We need to identify how similar a cell is to another cell, across every cell across these dimensions. For this, we will use the kNN, or k-nearest neighbor, graph, to identify which cells are close together and which are not. The kNN graph plots connections between cells if their distance (when plotted in this 20 dimensional space!) is amonst the k-th smallest distances from that cell to other cells. This will be crucial for identifying clusters, and is necessary for plotting a UMAP. [Larger neighbor values will result in more global structure being preserved at the loss of detailed local structure. In general this parameter should often be in the range 5 to 50, with a choice of 10 to 15 being a sensible default](https://github.com/lmcinnes/umap).
```
Comment (mtekman) [inline topics/transcriptomics/tutorials/scrna-seq-basic-pipeline/tutorial.md:547]: "we will use the kNN, or k-nearest neighbor, graph," → "we will use the k-nearest neighbor (kNN) graph" I think works better

File: topics/transcriptomics/tutorials/scrna-seq-basic-pipeline/tutorial.md [hunk 11]
```diff
@@ -0,0 +1,910 @@
...
+
+> ### {% icon details %} Working in a group? Decision-time!
+> If you are working in a group, you can now divvy up a decision here with one *control* and the rest varied numbers so that you can compare results throughout the tutorials.
```
Comment (mtekman) [inline topics/transcriptomics/tutorials/scrna-seq-basic-pipeline/tutorial.md:550]: "divvy up"

File: topics/transcriptomics/tutorials/scrna-seq-basic-pipeline/tutorial.md [hunk 12]
```diff
@@ -0,0 +1,910 @@
...
+>    - *"Maximum number of neighbours used"*: `15`
+>    - *"Number of PCs to use"*: `20`
+{: .hands_on}
+
+## Dimensionality reduction for visualisation
+
+Two major visualisations for this data are tSNE and UMAP. We must calculate the coordinates for both prior to visualisation. For tSNE, the parameter [**perplexity**](https://www.nature.com/articles/s41467-019-13056-x) can be changed to best represent the data, while for UMAP the main change would be to change the kNN graph above itself, by changing the **neighbours**.
+
+> ### {% icon details %} Working in a group? Decision-time!
+> If you are working in a group, you can now divvy up a decision here with one *control* and the rest varied numbers so that you can compare results throughout the tutorials.
```
Comment (mtekman) [inline topics/transcriptomics/tutorials/scrna-seq-basic-pipeline/tutorial.md:571]: DIVVY UP

File: topics/transcriptomics/tutorials/scrna-seq-basic-pipeline/tutorial.md [hunk 13]
```diff
@@ -0,0 +1,910 @@
...
+
+Ultimately, there are quite a lot ways to analyse the data, both within the confines of this tutorial (the many parameters that could be changed throughout) and outside of it (batch correction, sub-clustering, cell-cycle scoring, trajectories, etc.) Most analyses will still yield the same general output, though: there are fewer knockout cells in the mature T-cell population.
+
+{% icon congratulations %} Congratulations! You have interpreted your plots in several important ways!
+
+# Interactive visualisations
+
+Before we leave you to explore the unknown, you might have noticed that the above interpretations are only a few of the possible options. Plus you might have had fun trying to figure out which sample is which genotype is which sex and flicking back and forth between plots repeatedly. Figuring out which plots will be your *final publishable* plots takes a lot of time and testing. Luckily, there is a helpful interactive viewer export tool {% cite Moreno2020.04.08.032698 %} that can help you explore without having to produce new plots over and over!
+
+> ### {% icon hands_on %} Hands-on: Cellxgene
```
Comment (mtekman) [inline topics/transcriptomics/tutorials/scrna-seq-basic-pipeline/tutorial.md:874]: Nice reference to CellXGene here

File: snippets/create_new_workflow.md [hunk 14]
```diff
@@ -4,4 +4,5 @@
 > 2. Click the new workflow {% icon galaxy-wf-new %} button
 > 3. Give it a clear and memorable name
 > 4. Clicking **Save** will take you directly into the workflow editor for that workflow
-{: .tip}
\ No newline at end of file
+> 5. Need more help? Please see the *How to make a workflow subsection* [here](https://training.galaxyproject.org/training-material/topics/introduction/tutorials/galaxy-intro-101-everyone/tutorial.html)
```
Comment (hexylena) [inline snippets/create_new_workflow.md:7]:
```suggestion
> 5. Need more help? Please see the *How to make a workflow subsection* [here]({% link topics/introduction/tutorials/galaxy-intro-101-everyone/tutorial.md %})
```

File: topics/transcriptomics/tutorials/scrna-seq-basic-pipeline/workflows/main_workflow.ga [hunk 15]
```diff
@@ -0,0 +1,2561 @@
...
+            "workflow_outputs": [
+                {
+                    "label": null,
+                    "output_name": "output_png",
+                    "uuid": "180f366c-6007-4b0a-9c59-b05e2e4bb757"
+                }
+            ]
+        }
+    },
+    "tags": [],
```
Comment (mtekman) [inline topics/transcriptomics/tutorials/scrna-seq-basic-pipeline/workflows/main_workflow.ga:2558]: you need to add `"tags": ["transcriptomics"],` here

**Comment by hexylena** [comment]
It seems there are some linting comments in topics/transcriptomics/tutorials/scrna-seq-basic-pipeline/tutorial.md (you can check the 'files' tab to see some of them.)

**Review by mtekman — CHANGES_REQUESTED** [review]
Really comprehensive tutorial that covers a lot in a fun way

Some small typos, and one phrase that I am clearly triggered by, and I think it'd be good to be merged

File: topics/transcriptomics/tutorials/scrna-seq-basic-pipeline/tutorial.md (hunk 2 above)
Comment (hexylena) [reply topics/transcriptomics/tutorials/scrna-seq-basic-pipeline/tutorial.md:60]: @mtekman are you familiar with github's suggestion mode? It'd save @nomadscientist time when she goes to apply your review comments

Comment (mtekman) [reply topics/transcriptomics/tutorials/scrna-seq-basic-pipeline/tutorial.md:60]: oh whoops, I forgot about that...

Comment (hexylena) [reply topics/transcriptomics/tutorials/scrna-seq-basic-pipeline/tutorial.md:60]: went ahead and edited your feedback to do it :)

File: topics/transcriptomics/tutorials/scrna-seq-basic-pipeline/tutorial.md (hunk 1 above)
Comment (nomadscientist) [reply topics/transcriptomics/tutorials/scrna-seq-basic-pipeline/tutorial.md:42]: Done!

File: topics/transcriptomics/tutorials/scrna-seq-basic-pipeline/tutorial.md (hunk 13 above)
Comment (nomadscientist) [reply topics/transcriptomics/tutorials/scrna-seq-basic-pipeline/tutorial.md:874]: There's no actual cellxgene publication! I put in one that references it... (by its GitHub site)

**Comment by nomadscientist** [comment]
@mtekman @hexylena Thank you guys so much for your edits! Ok, I think it's all sorted now?

**Comment by hexylena** [comment]
there's some error with the frontmatter but i'm not sure what it is. I'll look into it.

**Comment by nomadscientist** [comment]
Yes good shout! OK sorted on that now

**Review by hexylena — APPROVED** [review]
I've pushed a few formatting fixes @nomadscientist but otherwise it looks good! (Check the commit log for what I did, but use 'snippet' now instead of 'include' is the main point)

**Comment by nomadscientist** [comment]
Ohhh @hexylena that datatypes bit is COOL! Thanks so much for your edits! I'll make sure to apply them to the next one in advance!

**Comment by hexylena** [comment]
Thanks so much @nomadscientist for a great tutorial (love how many question boxes are in there) and thanks @mtekman for reviewing it!

**Pull request merged**